collectors/ebpf.plugin/ebpf_vfs.c

```diff
@@ -0,0 +1,949 @@
+static char *vfs_id_names[NETDATA_KEY_PUBLISH_VFS_END] = { "vfs_unlink", "vfs_read", "vfs_write",
+                                                           "vfs_fsync", "vfs_open", "vfs_create"};
+
+static netdata_idx_t *vfs_hash_values = NULL;
+static netdata_syscall_stat_t vfs_aggregated_data[NETDATA_KEY_PUBLISH_PROCESS_END];
+static netdata_publish_syscall_t vfs_publish_aggregated[NETDATA_KEY_PUBLISH_PROCESS_END];
+netdata_publish_vfs_t **vfs_pid = NULL;
+netdata_publish_vfs_t *vfs_vector = NULL;
+
+static ebpf_data_t vfs_data;
+
+static ebpf_local_maps_t vfs_maps[] = {{.name = "tbl_vfs_pid", .internal_input = ND_EBPF_DEFAULT_PID_SIZE,
+                                        .user_input = 0},
+                                       {.name = NULL, .internal_input = 0, .user_input = 0}};
+
+struct config vfs_config = { .first_section = NULL,
+    .last_section = NULL,
+    .mutex = NETDATA_MUTEX_INITIALIZER,
+    .index = { .avl_tree = { .root = NULL, .compar = appconfig_section_compare },
+    .rwlock = AVL_LOCK_INITIALIZER } };
+
+static struct bpf_object *objects = NULL;
+static struct bpf_link **probe_links = NULL;
+
+struct netdata_static_thread vfs_threads = {"VFS KERNEL",
+                                            NULL, NULL, 1, NULL,
+                                            NULL,  NULL};
+
+static int *map_fd = NULL;
+
+static int read_thread_closed = 1;
+
+/*****************************************************************
+ *
+ *  FUNCTIONS TO CLOSE THE THREAD
+ *
+ *****************************************************************/
+
+/**
+ * Clean PID structures
+ *
+ * Clean the allocated structures.
+ */
+void clean_vfs_pid_structures() {
+    struct pid_stat *pids = root_of_pids;
+    while (pids) {
+        freez(vfs_pid[pids->pid]);
+
+        pids = pids->next;
+    }
+}
+
+/**
+* Clean up the main thread.
+*
+* @param ptr thread data.
+**/
+static void ebpf_vfs_cleanup(void *ptr)
+{
+    ebpf_module_t *em = (ebpf_module_t *)ptr;
+    if (!em->enabled)
+        return;
+
+    heartbeat_t hb;
+    heartbeat_init(&hb);
+    uint32_t tick = 50 * USEC_PER_MS;
+    while (!read_thread_closed) {
+        usec_t dt = heartbeat_next(&hb, tick);
+        UNUSED(dt);
+    }
+
+    freez(vfs_data.map_fd);
+    freez(vfs_hash_values);
+    freez(vfs_vector);
+
+    struct bpf_program *prog;
+    size_t i = 0 ;
+    bpf_object__for_each_program(prog, objects) {
+        bpf_link__destroy(probe_links[i]);
+        i++;
+    }
+    bpf_object__close(objects);
+}
+
+/*****************************************************************
+ *
+ *  FUNCTIONS WITH THE MAIN LOOP
+ *
+ *****************************************************************/
+
+/**
+ * Send data to Netdata calling auxiliar functions.
+ *
+ * @param em the structure with thread information
+*/
+static void ebpf_vfs_send_data(ebpf_module_t *em)
+{
+    netdata_publish_vfs_common_t pvc;
+
+    pvc.write = ((long)vfs_aggregated_data[NETDATA_KEY_PUBLISH_VFS_WRITE].bytes);
```

Comment:
minor - bit too many `()`

Reply:
It was removed with on of the commits I brought now.